restli-server/src/main/java/com/linkedin/restli/internal/server/model/RestLiAnnotationReader.java

```diff
@@ -2694,9 +2697,142 @@ else if (CollectionResult.class.isAssignableFrom(returnType))
           + resourceModel.getResourceClass().getName() + '\'');
     }
 
+    // Validate the linked batch finder if any for structural conformance.
+    if (linkedBatchFinderName != null)
+    {
+      validateLinkedBatchFinder(finderMethodDescriptor, resourceModel, linkedBatchFinderName);
+    }
+
     // query parameters are checked in getQueryParameters method
   }
 
+  private static void validateLinkedBatchFinder(ResourceMethodDescriptor finder,
+      ResourceModel resourceModel, String linkedBatchFinderName)
+  {
+    ResourceMethodDescriptor batchFinder =
+        resourceModel.findBatchFinderMethod(linkedBatchFinderName);
+
+    if (batchFinder == null)
+    {
+      throw new ResourceConfigException("Did not find any Linked @BatchFinder method named '"
+          + linkedBatchFinderName
+          + "' from @Finder Method named '" + finder.getFinderName()
+          + "' on class '"
+          + resourceModel.getResourceClass().getName() + '\'');
+    }
+
+    ParameterizedType collectionType = (ParameterizedType) getLogicalReturnType(batchFinder.getMethod());
+    final Class<?> batchFinderCriteriaType = (Class<?>) collectionType.getActualTypeArguments()[0];
+    final Class<?> batchFinderMetadataType = (Class<?>) collectionType.getActualTypeArguments()[2];
+    final Class<?> finderMetadataType = finder.getCollectionCustomMetadataType();
+    if (finderMetadataType != null && !finderMetadataType.equals(batchFinderMetadataType))
+    {
+      throw new ResourceConfigException("Linked @BatchFinder method '" + linkedBatchFinderName
+          + "' from @Finder Method named '" + finder.getFinderName()
+          + "' on class '" + resourceModel.getResourceClass().getName()
+          + "' does not have the same metadata type.");
+    }
+
+    final RecordDataSchema criteriaSchema = (RecordDataSchema) DataTemplateUtil.getSchema(batchFinderCriteriaType);
+    Set<String> fieldNames = criteriaSchema.getFields()
+        .stream()
+        .map(RecordDataSchema.Field::getName)
+        .collect(Collectors.toCollection(HashSet::new));
+    finder.getParameters().forEach(parameter ->
+    {
+      switch (parameter.getParamType())
+      {
+        case QUERY:
+        case ASSOC_KEY_PARAM:
```

Comment:
Since you include deprecated @Context here, you also should include deprecated @AssocKey.